### PR TITLE
[10.0][IMP] product_template_multi_link - Add link wizard

### PR DESCRIPTION
**Pull request by acsonefho**
Update module `product_templalte_multi_link` to add possibility to link multiple product in one shot.

- Go on product (template) tree view (or form view) and select products to link together.
- Click on Action menu, then "Manage Product Links".
- You can select the operation (remove every links or links together)
- Validate

### CODE REVIEW

**Review by Cedric-Pigeon — APPROVED** [review]

**Comment by HaraldPanten** [comment]
I would set this PR as an IMP. Wouldn't you?

File: product_template_multi_link/models/product_template_link.py [hunk 1]
```diff
@@ -5,16 +5,16 @@
 
 from odoo import fields, models
 
+_LINK_TYPE_SELECTION = [
```
Comment (simahawk) [inline product_template_multi_link/models/product_template_link.py]: pls move this to `_selection_link_type` method instead.

File: product_template_multi_link/tests/test_product_template_linker.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo.tests.common import TransactionCase
+
+
+class TestProductTemplateLinker(TransactionCase):
```
Comment (simahawk) [inline product_template_multi_link/tests/test_product_template_linker.py]: pls:

1. use SavepointCase
2. move `setUp` code to `setUpClass`
3. use `cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))` after calling super

:)

File: product_template_multi_link/wizards/product_template_linker.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+             "Link these products: will link all selected "
+             "products together.",
+    )
+    product_ids = fields.Many2many(
+        comodel_name="product.template",
+        string="Products",
+    )
+    link_type = fields.Selection(
+        string='Link Type',
+        selection=_LINK_TYPE_SELECTION,
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]:
```suggestion
        selection=lambda self: self.env['product.template.link']._selection_link_type()
```

File: product_template_multi_link/wizards/product_template_linker.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        if self.operation_type == "link":
+            self.action_apply_link()
+        elif self.operation_type == "unlink":
+            self.action_apply_unlink()
+        return {}
+
+    @api.multi
+    def action_apply_unlink(self):
+        """
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]:
```suggestion
        """Remove links from products.
```

File: product_template_multi_link/wizards/product_template_linker.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        :return: product.template.link recordset
+        """
+        self.product_ids.mapped("product_template_link_ids").unlink()
+        return self.env['product.template.link'].browse()
+
+    @api.multi
+    def action_apply_link(self):
+        """
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]:
```suggestion
        """Add link to products.
```

File: product_template_multi_link/wizards/product_template_linker.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            linked_products = existing_links.mapped(
+                "linked_product_template_id")
+            products_to_link = self.product_ids - linked_products - product
+            links |= self._create_link(product, products_to_link)
+        return links
+
+    @api.multi
+    def _create_link(self, product_source, target_products):
+        """
+        Create the link between given product source and target products.
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]:
```suggestion
        """Create the link between given product source and target products.
        
```

File: product_template_multi_link/wizards/product_template_linker.xml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+<!-- Copyright 2020 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record model="ir.ui.view" id="product_template_linker_form_view">
+        <field name="name">product.template.linker.form (in product_template_multi_link)</field>
+        <field name="model">product.template.linker</field>
+        <field name="arch" type="xml">
+            <form>
+                <group>
+                    <group>
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.xml]: name groups

**Review by simahawk — CHANGES_REQUESTED** [review]

File: product_template_multi_link/wizards/product_template_linker.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    @api.model
+    def default_get(self, fields_list):
+        """
+        Inherit default_get to auto-fill product_ids with current context
+        :param fields_list: list of str
+        :return: dict
+        """
+        result = super(ProductTemplateLinker, self).default_get(fields_list)
+        ctx = self.env.context
+        active_ids = ctx.get('active_ids', ctx.get('active_ids', []))
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]: :thinking:

File: product_template_multi_link/wizards/product_template_linker.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            linked_products = existing_links.mapped(
+                "linked_product_template_id")
+            products_to_link = self.product_ids - linked_products - product
+            links |= self._create_link(product, products_to_link)
+        return links
+
+    @api.multi
+    def _create_link(self, product_source, target_products):
+        """
+        Create the link between given product source and target products.
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.py]: the 1st line on docstrings must stay as per my comment: separated ;)
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: product_template_multi_link/wizards/product_template_linker.xml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        <field name="name">product.template.linker.form (in product_template_multi_link)</field>
+        <field name="model">product.template.linker</field>
+        <field name="arch" type="xml">
+            <form>
+                <group name="main_group">
+                    <group name="main_data_group">
+                        <field name="operation_type"/>
+                        <field name="link_type" attrs="{'invisible': [('operation_type', '!=', 'link')], 'required': [('operation_type', '=', 'link')]}"/>
+                    </group>
+                    <group></group>
```
Comment (simahawk) [inline product_template_multi_link/wizards/product_template_linker.xml]: empty group needed?

File: product_template_multi_link/wizards/product_template_linker.py (hunk 8 above)
Comment (acsonefho) [reply product_template_multi_link/wizards/product_template_linker.py]: Haha copy/paste.. It should be `active_id` in the second place :fearful:

**Pull request closed without merging**